crates/data_components/src/util/column_reference.rs

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use itertools::Itertools;
+use snafu::prelude::*;
+use std::{fmt::Display, hash::Hash};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display(r#"The column reference "{column_ref}" is missing a closing parenthensis."#))]
+    MissingClosingParenthesisInColumnReference { column_ref: String },
+}
+
+#[derive(Debug, Clone, Eq)]
+pub struct ColumnReference {
+    columns: Vec<String>,
+}
+
+impl ColumnReference {
+    #[must_use]
+    pub fn new(columns: Vec<String>) -> Self {
+        Self {
+            columns: columns.into_iter().sorted().collect(),
+        }
+    }
+
+    #[must_use]
+    pub fn empty() -> Self {
+        Self { columns: vec![] }
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = &str> {
+        self.columns.iter().map(String::as_str)
+    }
+
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.columns.is_empty()
+    }
+
+    #[must_use]
+    pub fn contains(&self, column: &String) -> bool {
+        self.columns.contains(column)
+    }
+}
+
+impl Default for ColumnReference {
+    fn default() -> Self {
+        Self::empty()
+    }
+}
+
+impl PartialEq for ColumnReference {
+    fn eq(&self, other: &Self) -> bool {
+        if self.columns.len() != other.columns.len() {
+            return false;
+        }
+
+        self.columns
+            .iter()
+            .zip(other.columns.iter())
+            .all(|(a, b)| a == b)
+    }
+}
+
+impl Hash for ColumnReference {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.columns.hash(state);
+    }
+}
+
+/// Parses column references from a string into a vector of each individual column reference.
```

Comment:
This is a new struct for this PR. This represents a logical reference to a set of columns, with some helpers for serialization/de-serialization to/from strings.

These column references are the same as used in the `primary_key` and `indexes` configuration